trieste/ask_tell_optimization.py

```diff
@@ -0,0 +1,251 @@
+# Copyright 2021 The Trieste Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module contains the Ask/Tell API for users of Trieste who would like to
+perform Bayesian Optimization with external control of the optimization loop.
+"""
+
+
+from __future__ import annotations
+
+from typing import Dict, Generic, Mapping, Optional, TypeVar, cast, overload
+
+from .acquisition.rule import AcquisitionRule, EfficientGlobalOptimization
+from .bayesian_optimizer import OptimizationResult, Record
+from .data import Dataset
+from .models import ModelSpec, create_model
+from .observer import OBJECTIVE
+from .space import SearchSpace
+from .types import State, TensorType
+from .utils import Ok, map_values
+
+S = TypeVar("S")
+""" Unbound type variable. """
+
+SP = TypeVar("SP", bound=SearchSpace)
```

Comment:
this can be imported from `space`, right?

Reply:
I agree, we are not saving much here and it seems to be harmless duplication. In any case, I'm still learning about typing, so you are probably right.